### ISSUE TITLE
Research issues related to getblocktemplate timings with shielded coinbase

### ISSUE THREAD

**Issue by steven-ecc** [zcash/zcash#4817]

**Comment by LarryRuane** [comment]
@nitronick600, what about a variant of `getblocktemplate` that differs (from `getblocktemplate`) in two ways:
- takes a height argument
- returns an empty block (i.e. does not include any transactions from the mempool)

The coinbase tx would be build from the `-mineraddress` configuration variable. This should be predictably fast. A good name might be `getemptyblocktemplate`.

**Comment by nitronick600** [comment]
@LarryRuane I like the idea of a `getemptyblocktemplate` RPC in theory, but in practice it may be abused - It's much easier to mine empty blocks.  For a chain like ZCash in its infancy, there's not much tx volume to incentivize mining anything but empty blocks. It'd be good to get buy-in from ECC on that. 
 
I'm also not a huge fan of `-mineraddress` because it requires restarting the daemon which can be very slow.  It's handy to be able to swap out the miner address on the fly.

**Comment by LarryRuane** [comment]
I may have just had an idea in which we don't need any new RPC. (Ignore my previous comment.) Are you using `getblocktemplate` with the `longpollid` argument? What arguments are you using exactly?

The way `getblocktemplate` currently works is, it waits for the `longpollid` event (if specified) to occur (either a new block, or the mempool transaction count changes), and _then_ it computes the block template, including the coinbase transaction (which is somewhat slow if mining to a shielded address).

What if, instead, when that RPC _begins_ running, it _immediately_ computes the coinbase transaction, and _then_ it waits for the `longpollid` event; when it occurs, the RPC can return much more quickly since it's computed the coinbase transaction already.

This wouldn't require any interface changes; it would be strictly a performance improvement.

Would this work?

**Comment by LarryRuane** [comment]
Minor correction:

> This wouldn't require any interface changes

Probably as a separate PR (than the one to improve performance), we would still want to add an optional `mineraddress` argument (which could be either a transparent or shielded address):
```
zcash-cli getblocktemplate '{"mineraddress":"(addr)"}'
```
If not specified, the behavior would be as it is [currently](https://github.com/zcash/zcash/blob/877212414ad40e37cf8e49884a90767c54d59ba2/src/wallet/wallet.cpp#L4413), i.e., it would use the `-mineraddress=` command-line argument (as today), and if that's not specified, we'd create a new transparent address.

Small decision to be made: Currently, as a safety measure, the mining address given by `-mineraddress` must be part of the local wallet; this can be overridden with `-minetolocalwallet=0`. Should this safety check also apply when the mining address is specified in the `getblocktemplate` rpc?

**Comment by nitronick600** [comment]
` Are you using getblocktemplate with the longpollid argument?` We don't use longpollid - should we be?

`this can be overridden with -minetolocalwallet=0` IMO this isn' necessary.  Anyone calling `getblocktemplate` has a fundamental understanding of the security implications. 

The only downside I see here is the consensus change case.  Currently, we pre-compute an empty block w/ a coinbase transaction that we can create in memory - this takes 10-15ms.  How can we ensure that performance is maintained?

**Comment by LarryRuane** [comment]
> We don't use longpollid - should we be?

I would say so, yes. When a new block arrives, how exactly does your stratum server discover that's happened? Are you using the `-blocknotify` mechanism (in the configuration file or command-line)? I'll assume that's the case -- then does the program that runs when a new block arrives (the argument to `-blocknotify`) submit your precomputed empty block? But you can't have precomputed the entire block, because its prevhash depends on the block that just arrived. So I assume you probably quickly construct an empty block with the precomputed coinbase transaction? Then, after the miners have been started on this new block, you've got a little breathing room so maybe you a `getblocktemplate` RPC (to get a bunch of transactions)? 

If all of that is correct, and if this is working for you, is there a reason you'd like us to make changes?

The `longpollid` feature of `getblocktemplate` is very cool (once my PR, #4886, is merged), and will at least simplify what you're doing now (if I understand correctly what you're doing now), and probably will offer some performance advantage.

The way this feature works is, you submit this `longpollid` argument with the `getblocktemplate` call, and the RPC will _block_ until a new block arrives. When that happens, it will return _immediately_ with a block template based on this new block, and you can use that to start mining immediately. PR #4886 precomputes the coinbase (only if you're mining to a shielded address), and returns an empty block -- there's no way around that (other than making the calculation of a shielded output faster, which may not even be possible; this PR doesn't try to improve this). The return from `getblocktemplate` is within 1 or two milliseconds. (Today, without #4886, it doesn't return for about 400 ms on my laptop.) So this seems like it would be perfect for you (and anyone mining to a shielded address).

It's not as bad as you'd think to return an empty block, because the block that just arrived most likely swept up the entire mempool, so you can only attempt to mine an empty block for at least some seconds after a new block arrives. When blocks become full, this idea of returning an empty block would be more of a problem. But even then, not a huge problem, because the other thing `longpollid` does is, it checks the mempool every 10 seconds, and if there are new transactions (or, if attempting to mine an empty block, _any_ transactions), then it returns a block template with all of those transactions (or as many as will fit; currently, that all). This does require computing a shielded coinbase, but that's okay because a new block hasn't just arrived (so your miners are still doing useful hashing). You can then have them switch to this new version of the block that has a bunch of transactions, which I assume can happen quickly (but that you wouldn't want to do more often than every 10 seconds or so).

So the only time an actual empty block gets mined is if two blocks arrive close together, within about 10 seconds. But that can happen even today when mining to a transparent address -- the first of the two close-together blocks clears out the mempool (because blocks aren't full), and the second has no transactions to include.

The value of the `longpollid` argument is should be the `longpollid` from the previous result. (The first time, just don't specify `longpollid`; the RPC will return immediately, and the response will include a `longpollid` string to use for the next call.)
